server-utils/src/cors.rs

```diff
@@ -0,0 +1,594 @@
+//! CORS handling utility functions
+use unicase;
+
+pub use self::unicase::Ascii;
+use crate::hosts::{Host, Port};
+use crate::matcher::{Matcher, Pattern};
+use std::collections::HashSet;
+use std::{fmt, ops};
+
+/// Origin Protocol
+#[derive(Clone, Hash, Debug, PartialEq, Eq)]
+pub enum OriginProtocol {
+	/// Http protocol
+	Http,
+	/// Https protocol
+	Https,
+	/// Custom protocol
+	Custom(String),
+}
+
+/// Request Origin
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct Origin {
+	protocol: OriginProtocol,
+	host: Host,
+	as_string: String,
+	matcher: Matcher,
+}
+
+impl<T: AsRef<str>> From<T> for Origin {
+	fn from(string: T) -> Self {
+		Origin::parse(string.as_ref())
+	}
+}
+
+impl Origin {
+	fn with_host(protocol: OriginProtocol, host: Host) -> Self {
+		let string = Self::to_string(&protocol, &host);
+		let matcher = Matcher::new(&string);
+
+		Origin {
+			protocol,
+			host,
+			as_string: string,
+			matcher,
+		}
+	}
+
+	/// Creates new origin given protocol, hostname and port parts.
+	/// Pre-processes input data if necessary.
+	pub fn new<T: Into<Port>>(protocol: OriginProtocol, host: &str, port: T) -> Self {
+		Self::with_host(protocol, Host::new(host, port))
+	}
+
+	/// Attempts to parse given string as a `Origin`.
+	/// NOTE: This method always succeeds and falls back to sensible defaults.
```

Comment:
That was a decision we made for backward compatibiltiy afair, but I think in `jsonrpsee` we could conside enforcing a proper format of `Origin`.

I.e. now we default to `http` if no protocol is given, because the `Origin` header needs a protocol, instead we could just fail with error and force user to specify that.

@tomaka your call which direction you would prefer this project to go.